src/util/find-root.ts

```diff
@@ -0,0 +1,179 @@
+
+      const pkg = await safeReadJson<PJSON>(join(next, 'package.json'))
+      if (pkg?.name === name) return next
+    } else {
+      cur = join(next, 'package.json')
+      if (await safeReadJson<PJSON>(cur)) return dirname(cur)
+    }
+  }
+}
+
+let pnp: {
+  getPackageInformation: typeof getPackageInformation
+  getLocator: (name: string, reference: string | [string, string]) => PackageLocator
+  getDependencyTreeRoots: () => PackageLocator[]
+}
+
+/**
+ * The pnpapi module is only available if running in a pnp environment. Because of that
+ * we have to require it from the plugin.
+ *
+ * Solution taken from here: https://github.com/yarnpkg/berry/issues/1467#issuecomment-642869600
+ */
+function maybeRequirePnpApi(root: string): unknown {
+  if (pnp) return pnp
+  try {
+    // eslint-disable-next-line node/no-missing-require
+    pnp = require(require.resolve('pnpapi', {paths: [root]}))
+    return pnp
+  } catch {}
+}
+
+const getKey = (locator: PackageLocator | string | [string, string] | undefined) => JSON.stringify(locator)
+const isPeerDependency = (
+  pkg: PackageInformation | undefined,
+  parentPkg: PackageInformation | undefined,
+  name: string,
+) => getKey(pkg?.packageDependencies.get(name)) === getKey(parentPkg?.packageDependencies.get(name))
+
+/**
+ * Traverse PnP dependency tree to find plugin root directory.
+ *
+ * Implementation adapted from https://yarnpkg.com/advanced/pnpapi#traversing-the-dependency-tree
+ */
+function findPnpRoot(name: string, root: string): string | undefined {
+  maybeRequirePnpApi(root)
+
+  if (!pnp) return
+
+  const seen = new Set()
+
+  const traverseDependencyTree = (locator: PackageLocator, parentPkg?: PackageInformation): string | undefined => {
+    // Prevent infinite recursion when A depends on B which depends on A
+    const key = getKey(locator)
+    if (seen.has(key)) return
+
+    const pkg = pnp.getPackageInformation(locator)
+
+    if (locator.name === name) {
+      return pkg.packageLocation
+    }
+
+    seen.add(key)
+
+    for (const [name, referencish] of pkg.packageDependencies) {
+      // Unmet peer dependencies
+      if (referencish === null) continue
+
+      // Avoid iterating on peer dependencies - very expensive
+      if (parentPkg !== null && isPeerDependency(pkg, parentPkg, name)) continue
+
+      const childLocator = pnp.getLocator(name, referencish)
+      const foundSomething = traverseDependencyTree(childLocator, pkg)
+      if (foundSomething) return foundSomething
+    }
+
+    // Important: This `delete` here causes the traversal to go over nodes even
+    // if they have already been traversed in another branch. If you don't need
+    // that, remove this line for a hefty speed increase.
+    seen.delete(key)
+  }
+
+  // Iterate on each workspace
+  for (const locator of pnp.getDependencyTreeRoots()) {
+    const foundSomething = traverseDependencyTree(locator)
+    if (foundSomething) return foundSomething
+  }
```

Comment:
This is what I referenced with

> If there are multiple versions of the plugin that's trying to be found, the wrong one might be selected (root is ignored)

In https://github.com/oclif/core/issues/806

And I'm not sure if the issue is fixed?

Suppose that someone has two workspaces and they each contain a different version of the oclif plugin, this wouldn't necessarily take the oclif plugin in the correct workspace, but rather the first one found. This might lead to frustrating issues where someone updates their plugin version (in the workspace they have it installed in) but it doesn't do anything because oclif picks a workspace that has an older version.

Reply:
@danieltroger This might be a scenario where I prefer to wait until someone actually has this issue. Also I see this code path as a fallback for when people do not specify a `main` or `exports`, which is something we strongly advise against because it's _much_ more performant to do a `require.resolve` than it is to traverse the dependency tree for a matching plugin.

So if someone were to have such an issue, I would first suggest that they define `main` or `exports` in their package.json.